### PR TITLE
Make possible to reference previous pillars from subsequent pillars, as they specified in the top file

### PR DESCRIPTION
**Pull request by AndreiPashkin**
### What does this PR do?

It implements [proposal](https://github.com/saltstack/salt/issues/6955#issuecomment-45918224) of @jberkus to make possible to reference previous pillars from the ones that are subsequent in the top file.

Even if core developers have own plans for solving #6955, this feature does not prevent any other solution from being implemented and would just give a developer another way to set up his system.

### What issues does this PR fix or reference?
#6955
### Previous Behavior

Previously order in which regular pillars are loaded (and become reference-able by other pillars) was undefined. 
### New Behavior

Now user can reference pillars, that come before current pillar in the top file.
### Tests written?

Yes. I would ask core devs to advise if and what additional tests are needed.

### CODE REVIEW

**Comment by cachedout** [comment]
I would like to have @thatch45 @terminalmage and @plastikos all chime in here.

**Comment by thatch45** [comment]
I like this approach and I am ok with it. It inserts the logic in the correct place and simplifies the code a little, I think it is very well done.

**Comment by AndrewPashkin** [comment]
The PR is not yet ready to be merged, wait until I finalize it plz.

**Comment by thatch45** [comment]
Of course :)

**Comment by AndrewPashkin** [comment]
I'm trying to fix the tests, but when I launch them locally - they pass (unlike in Jenkins). For example this one:
https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu14-n/6299/testReport/integration.modules.state/StateModuleTest/test_onfail_requisite/

Can somebody tell, from the top of the head, what can cause this?

**Comment by AndrewPashkin** [comment]
Okay, I have updated `GitPython` from `1.X` to `2.X` and the test failed in my local environment, I'll try to handle it by myself.

**Comment by AndrewPashkin** [comment]
I resolved the issues with tests, but stuck with #37905.

**Comment by AndrewPashkin** [comment]
@cachedout, hooray, the PR is ready to be merged now!

File: doc/topics/pillar/index.rst [hunk 1]
```diff
@@ -236,6 +236,51 @@ The resulting pillar will be as follows:
     For example, in the above scenario conflicting key values in ``services``
     will overwrite those in ``packages`` because it's at the bottom of the list.
 
+
+Referencing Other Pillars
+=========================
+
+.. versionadded:: ?????
```
Comment (terminalmage) [inline doc/topics/pillar/index.rst]: Now that we're past the Carbon release cycle, this can be changed to ``.. versionadded:: Nitrogen``

File: doc/topics/pillar/index.rst [hunk 2]
```diff
@@ -236,6 +236,51 @@ The resulting pillar will be as follows:
...
+
+It is possible to reference pillar values, that are defined in other
```
Comment (terminalmage) [inline doc/topics/pillar/index.rst]: The comma here is unnecessary.

File: doc/topics/pillar/index.rst [hunk 3]
```diff
@@ -236,6 +236,51 @@ The resulting pillar will be as follows:
...
+pillar files. To do that, the pillar with referenced pillars should
```
Comment (terminalmage) [inline doc/topics/pillar/index.rst]: This could be worded better. Perhaps:

> To do this, make sure that any pillar variables which are being referenced by other pillar SLS files are defined higher up in the pillar Top file, so that they are evaluated first.

**Review by terminalmage — COMMENTED** [review]
Some comments about the docs, doesn't look like it would block a merge, we can take care of them afterwards.

Code looks OK, the only thing that has me worried about performance is the invocations of the loader in ``render_pillar()``, this might put extra load on the master at scale when compiling pillar data for a large number of minions.

@thatch45 is this a valid concern, or am I just being too cautious?

File: doc/topics/pillar/index.rst (hunk 1 above)
Comment (AndrewPashkin) [inline doc/topics/pillar/index.rst]: Done.

File: doc/topics/pillar/index.rst (hunk 2 above)
Comment (AndrewPashkin) [inline doc/topics/pillar/index.rst]: Fixed.

File: doc/topics/pillar/index.rst (hunk 3 above)
Comment (AndrewPashkin) [inline doc/topics/pillar/index.rst]: Fixed.

**Comment by AndrewPashkin** [comment]
@terminalmage 
> the only thing that has me worried about performance is the invocations of the loader in render_pillar(), this might put extra load on the master at scale when compiling pillar data for a large number of minions.

What solutions do you see? To make loaders copy-able and mutable?

**Comment by cachedout** [comment]
That concern is very valid. Given that this is in a loop, this has the potential to totally destroy master performance at scale. Why not use the copy of the minion functions we already have in the pillar instance?

**Comment by AndrewPashkin** [comment]
> Why not use the copy of the minion functions we already have in the pillar instance?

Because without that `salt['pillar.get'](...)` won't be able to retrieve updated pillar values (from other pillar SLS).

**Comment by cachedout** [comment]
I'm not following why that would be. This looks like we're just using the recomputed functions to generate new renderers. Forgive my lack of understanding here but what's different about the renderers on each iteration?

**Comment by AndrewPashkin** [comment]
Take a look at this integration test:
https://github.com/saltstack/salt/pull/37003/files#diff-6ee6bddf72661e11ee47ae68054d55b2R127

And at how pillars are rendered for it:
https://github.com/saltstack/salt/pull/37003/files#diff-84dcf1edfcafc9be25c02685b9631767R2

Without recreating functions `uppercase_knights` would not be computed correctly, because `salt['pillar.get']('knights')` would return `None` (or something like that). Because functions load would not have recomputed pillars within it and so `pillar.get` function, while looking up `__pillar__` dict wold find old version of pillars.

But `lowercase_knights` would be computed correctly, because pillars are recreated each iteration and passed down to a renderer.

TL;DR To make functions like `pillar.get` and others "see" updated pillar values from previous iterations over SLS files, we need to recreate a function loader with passing new pillars to it.

**Comment by cachedout** [comment]
K, that makes sense. Thanks for the explanation.

**Comment by AndrewPashkin** [comment]
I think the solution is to add `update()` method to `LazyLoader`, that would make possible to update a bound data of the instance.

**Comment by cachedout** [comment]
@AndrewPashkin Yes, I agree. Exactly what I was thinking. Do you want to take a swing at that? Might be a bit before I can get to it.

**Comment by AndrewPashkin** [comment]
@cachedout, I'd want to do that, but I have bad shortage of time, so I can't really promise anything.

**Comment by AndrewPashkin** [comment]
@cachedout, @terminalmage,

Hi, guys, I decided to get back to this PR.

Here is the quick summary of the current state of things:
- Everything is working.
- But there are [concerns about performance][1]: within `render_pillar()` there are [construction][2] of `LazyLoader`, which loads the modules. This patch makes them be recreated in a loop where each iteration is the processing of each pillar SLS. So it makes modules to be reloaded for rendering of each SLS which is not good.
- We came up with the idea of making `LazyLoader` mutable, so instead of recreating it it would be possible to update it.

*BUT*. I've found that it seems like it is fundamentally impossible to make them updateable, because there is [the place][3] when the options (the thing that is actually needs to be made updateable in `LazyLoader`) are passed to an external module's initializer. So if we are passing the options to an external module and it does everything it wants with it we should re-initialize it in case if we are updated the options. It means that we can not escape the performance penalty and this is not a programming limitation, but rather a fundamental one.

What are your thoughts on that?

[1]: https://github.com/saltstack/salt/pull/37003#pullrequestreview-11245486
[2]: https://github.com/saltstack/salt/pull/37003/files#diff-91025f372e58dc0e88b6c7255c9da871R689
[3]: https://github.com/saltstack/salt/blob/38e179388365c2bcef87b32285d0f39f80b60094/salt/loader.py#L1397

**Comment by AndrewPashkin** [comment]
The only idea that I have for now is that it's possible to factor out actual loading of the modules out of `LazyLoader._load_module()` and make this function/method results cached. And so `LazyLoader` recreation will become less costly.

**Comment by AndrewPashkin** [comment]
@cachedout, @terminalmage,

I think, that it might be the case, that the performance impact might be not that significant. 

The bottle neck here is `LazyLoader._load_module()` - it performs actual loading of a Python module from disk. Before this change modules used in the SLS files were cached between rendering of the SLS files. Now they are reloaded each time if they are called in an SLS file. 

But considering that the modules are loaded lazily, on demand and that the amount of SLS files is not something that scales to really big numbers usually (like number of minions), the actual overhead might be not noticeable. 

This is also supported by the fact that test runs times are not increased (both ~1:15), and I also did local benchmarks with multiple minions and multiple SLS files invoking the same module and I haven't found any difference in performance between `develop` and my patch.

**Comment by cachedout** [comment]
Hmm, I think you might be right here.

@terminalmage and @thatch45 what do you think?

**Comment by AndrewPashkin** [comment]
That's correct, if a module defines `__init__` function it is [called][1] after loading.
Regarding caching of modules - how it's possible if modules are [mutated][2] to set globals like `__pillar__`, `__salt__`, etc?

[1]: https://github.com/saltstack/salt/blob/899fb6fc94acb72e9030e3ed6512f2e6714f9668/salt/loader.py#L1451
[2]: https://github.com/saltstack/salt/blob/899fb6fc94acb72e9030e3ed6512f2e6714f9668/salt/loader.py#L1444

**Comment by thatch45** [comment]
this is why we reload pillar modules with each request to the pillar, this was actually a bug in much earlier versions of salt where the cached modules did not always properly contain the right pillar data. So you are spot on once again

**Comment by AndrewPashkin** [comment]
I see that the Python 3 tests are failing and not only in my PR. Should I wait until they are fixed?

**Comment by AndrewPashkin** [comment]
Tests are fixed, let's :ship: it?

**Comment by AndrewPashkin** [comment]
Guys, I noticed a problem, in this [place][1], `functions` that are passed in the [constructor][2] are ignored, which is not right. Please, **revert** this PR.

I see the way to fix it - remove `functions` parameter from Pillar interface, which would affect `[Pillar][4]`, `[PillarCache][5]`, `[RemotePillar][6]`, `[RemotePillar][6]`, `[AsyncPillar][7]`, `[AsyncRemotePillar][8]` classes as well as the factories `[get_pillar][9]`, `[get_async_pillar][10]`.

@cachedout, @terminalmage, @thatch45, what are your thoughts on this? Are there any fundamental objections against removing `functions` parameter? 

The only place where this parameter is actually used is [here][3] and I think it could be refactored to not use it.

[1]: https://github.com/saltstack/salt/pull/37003/files#diff-91025f372e58dc0e88b6c7255c9da871R755
[2]: https://github.com/saltstack/salt/pull/37003/files#diff-91025f372e58dc0e88b6c7255c9da871L289
[3]: https://github.com/saltstack/salt/blob/e512a29f08a86f79ca91e9bf5036be3c84572ae5/salt/daemons/masterapi.py#L716
[4]: https://github.com/saltstack/salt/blob/7d59f7a6c60017190d262f3affdd445a132e4761/salt/pillar/__init__.py#L267
[5]: https://github.com/saltstack/salt/blob/7d59f7a6c60017190d262f3affdd445a132e4761/salt/pillar/__init__.py#L189
[6]: https://github.com/saltstack/salt/blob/7d59f7a6c60017190d262f3affdd445a132e4761/salt/pillar/__init__.py#L133
[7]: https://github.com/saltstack/salt/blob/7d59f7a6c60017190d262f3affdd445a132e4761/salt/pillar/__init__.py#L987
[8]: https://github.com/saltstack/salt/blob/7d59f7a6c60017190d262f3affdd445a132e4761/salt/pillar/__init__.py#L82
[9]: https://github.com/saltstack/salt/blob/7d59f7a6c60017190d262f3affdd445a132e4761/salt/pillar/__init__.py#L38
[10]: https://github.com/saltstack/salt/blob/7d59f7a6c60017190d262f3affdd445a132e4761/salt/pillar/__init__.py#L62
